common/src/main/java/org/keycloak/common/util/Environment.java

```diff
@@ -35,4 +38,16 @@ public static int getServerStartupTimeout() {
         }
     }
 
+    /**
+     * Tries to detect if Java platform is in the FIPS mode
+     * @return true if java is FIPS mode
+     */
+    public static boolean isJavaInFipsMode() {
+        for (Provider provider : Security.getProviders()) {
+            if (provider.getName().equals("BCFIPS")) continue; // Ignore BCFIPS provider for the detection as we may register it programatically
+            if (provider.getName().toUpperCase().contains("FIPS")) return true;
+        }
+        return false;
+    }
```

Comment:
Could you elaborate on this method? What other security providers has FIPS in it that are not BCFIPS?

Reply:
This is an attempt to "autodetect" if we are in the FIPS enabled platform or not.

IMO in the future, it may be good that administrator has the possibility to enable FIPS manually when running Keycloak. However at the same time, it can be good to "autodetect" the default in case that it is not overriden by administrator, so for example when Keycloak is executed on the FIPS enabled platform, Keycloak will automatically choose (probably at the time of running `./kc.sh build`) whether to use FIPS based crypto (including BCFIPS dependencies) or non-FIPS providers (like plain BC dependencies). So far, this "autodetection" is used only in the unit tests in this PR.

The autodetection is based on available Java providers. I am not aware of any better way right now. On the RHEL 8.6 with FIPS enabled, the default java security file has enabled some FIPS providers when running in the FIPS mode. The list of providers in the java.security file looks like this:
```
security.provider.1=SUN
security.provider.2=SunRsaSign
security.provider.3=SunEC
security.provider.4=SunJSSE
security.provider.5=SunJCE
security.provider.6=SunJGSS
security.provider.7=SunSASL
security.provider.8=XMLDSig
security.provider.9=SunPCSC
security.provider.10=JdkLDAP
security.provider.11=JdkSASL
security.provider.12=SunPKCS11
#security.provider.1=SunPKCS11 ${java.home}/lib/security/nss.cfg

#
# Security providers used when global crypto-policies are set to FIPS.
#
fips.provider.1=SunPKCS11 ${java.home}/conf/security/nss.fips.cfg
fips.provider.2=SUN
fips.provider.3=SunEC
fips.provider.4=com.sun.net.ssl.internal.ssl.Provider SunPKCS11-NSS-FIPS
```

The providers with the `security.provider` prefix are used just when Java application is executed in the non-fips mode with `-Dcom.redhat.fips=false` (very similar providers are also in the java file on my non-FIPS based Ubuntu laptop). The providers with `fips.provider` prefix are used by default on FIPS enabled RHEL 8.6 system.

Reply:
Feels a little bit flaky, but since it's only used by the tests it works for now.